modules/core/src/main/java/org/apache/ignite/internal/util/PointerWrapping.java

```diff
@@ -17,19 +17,268 @@
 
 package org.apache.ignite.internal.util;
 
+import static org.apache.ignite.internal.util.GridUnsafe.NATIVE_BYTE_ORDER;
+import static org.apache.ignite.internal.util.GridUnsafe.UNSAFE;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
 import java.nio.ByteBuffer;
+import java.security.AccessController;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
 
 /**
  * Wraps a pointer to unmanaged memory into a direct byte buffer.
  */
-@SuppressWarnings("InterfaceMayBeAnnotatedFunctional")
-interface PointerWrapping {
+abstract class PointerWrapping {
+    /** Null object. */
+    private static final Object NULL_OBJ = null;
+
+    private static final MethodHandle DIRECT_BUF_MTD;
+    private static final MethodHandle DIRECT_BUF_CTOR_INT;
+    private static final MethodHandle DIRECT_BUF_CTOR_LONG;
+
+    private static final Object JAVA_NIO_ACCESS_OBJ;
+
+    static {
+        Object nioAccessObj = null;
+
+        MethodHandle directBufMtd = null;
+        MethodHandle directBufCtorWithIntLen = null;
+        MethodHandle directBufCtorWithLongLen = null;
+
+        try {
+            directBufCtorWithIntLen = createAndTestNewDirectBufferCtor(int.class);
+        } catch (Exception e) {
+            try {
+                directBufCtorWithLongLen = createAndTestNewDirectBufferCtor(long.class);
+            } catch (Exception e2) {
+                try {
+                    nioAccessObj = javaNioAccessObject();
+                    directBufMtd = newDirectBufferMethodHandle(nioAccessObj);
+                } catch (Exception exFallback) {
+                    //noinspection CallToPrintStackTrace
+                    exFallback.printStackTrace(); // NOPMD
+
+                    e.addSuppressed(exFallback);
+
+                    throw e; // Fallback to shared secrets failed.
+                }
+
+                if (nioAccessObj == null || directBufMtd == null) {
+                    throw e;
+                }
+            }
+        }
+
+        DIRECT_BUF_MTD = directBufMtd;
+        DIRECT_BUF_CTOR_INT = directBufCtorWithIntLen;
+        DIRECT_BUF_CTOR_LONG = directBufCtorWithLongLen;
+
+        JAVA_NIO_ACCESS_OBJ = nioAccessObj;
+    }
+
     /**
      * Wraps a pointer to unmanaged memory into a direct byte buffer.
      *
      * @param ptr Pointer to wrap.
      * @param len Memory location length.
      * @return Byte buffer wrapping the given memory.
      */
-    ByteBuffer wrapPointer(long ptr, int len);
+    static ByteBuffer wrapPointer(long ptr, int len) {
```

Comment:
Can we implement this as a reference to a method that will be computed in the static block? Or will it bring the same performance penalties?

Reply:
I don't think I understood your proposal

Reply:
have a static field, like `private static final Function<Long, Int, ByteBuffer>` and assign a correct method reference inside the static block

Reply:
I see now. `static` with the combination of `if` makes the invoke of target MH completely static, which is the most optimal way of doing it. I'll add a comment

Reply:
> Don't do it without having benchmarks and guaranteeing that performance won't suffer.

Having this sentence without having the initial benchmarks yourself is kind of hypocritical

Reply:
I am in the process of committing that benchmark, don't worry

Reply:
Results are not as good as I saw in naive "for loop" tests, unfortunately

Reply:
Results are in Jira